Ability to restrict Emitter/Receiver channels

Adman opened the pull request:
* closes 2826

Signed-off-by: Adrian Matejov <ado.matejov@gmail.com>

**Description**
This feature allows to specify allowed channels for Emitter and Receiver nodes. When calling `wb_emitter_set_channel` or `wb_receiver_set_channel`, the channel is checked against allowed channels.

**Related Issues**
This pull-request fixes issue #2826 

**Tasks**
  - [x] Check whether initial channel is allowed during initialization
  - [x] Implement `updateAllowedChannels`

**Documentation**
https://www.cyberbotics.com/doc/reference/emitter
https://www.cyberbotics.com/doc/reference/receiver

**Additional context**
There are some small things that I don't know how or where to implement, so I would appreciate some advice on those :)
I'm also not that good in C++, so there might be some non-compilable mistakes there.

omichel commented:
Thank you for this contribution.
I believe you just need to insert the allowed channel information to the configure message to the Emitter and Receiver devices that are sent initially from Webots to the libController, see https://github.com/cyberbotics/webots/blob/develop/src/webots/nodes/WbEmitter.cpp#L112-L124 and https://github.com/cyberbotics/webots/blob/develop/src/controller/c/emitter.c#L128-L136 (and same for the Receiver). Let me know if this is unclear.

Adman commented:
@omichel Thanks. I introduced `allowedChannels_size` field in C files to actually store the the size of `allowedChannels` array -- the stream message now contains this size followed by allowed channels.

I'd like to ask about a couple of things:
1.What does this line mean? `connect(mAllowedChannels, &WbMFInt32::changed, this, &WbEmitter::updateAllowedChannels);`. Should I implement the function `updateAllowedChannels`? When is it called?

2. Where to implement initial check, whether the `channel` specified in the world file is actually allowed?

omichel commented:
> 1.What does this line mean? `connect(mAllowedChannels, &WbMFInt32::changed, this, &WbEmitter::updateAllowedChannels);`. Should I implement the function `updateAllowedChannels`? When is it called?

It will call the `updateAllowedChannels` method each time the `allowedChannels` is modified, either from the scene tree GUI or by an API call.

> 2. Where to implement initial check, whether the `channel` specified in the world file is actually allowed?

You should probably implement this in the `updateAllowedChannels`, see for example [this](https://github.com/cyberbotics/webots/blob/bf3cbb1b0a04df8ba4265ceb7f5d7632749b9031/src/webots/nodes/WbContactProperties.cpp#L86-L106).

Adman commented:
@omichel Thanks for the suggestion. I implemented it and successfully managed to compile the code and test it.

There is one small thing. What to do, when the world file contains following:
```
channel 1
allowedChannels [2]
```
Is it possible to raise something like "error in world definition", which would prevent running the controllers?

omichel commented on src/webots/nodes/WbEmitter.cpp:
```diff
@@ -105,7 +107,31 @@ void WbEmitter::updateRange() {
   mNeedToSetRange = true;
 }
 
+bool WbEmitter::is_channel_allowed() {
```
Method names should written in [lower camel case](https://github.com/cyberbotics/webots/wiki/CPP-Coding-Style#cs306-names-representing-methods-or-functions-must-be-written-in-lower-camel-case)
```suggestion
bool WbEmitter::isChannelAllowed() {
```

omichel commented on src/webots/nodes/WbEmitter.cpp:
```diff
@@ -105,7 +107,31 @@ void WbEmitter::updateRange() {
 }
 
+bool WbEmitter::is_channel_allowed() {
+  const int allowedChannels_size = mAllowedChannels->size();
```
```suggestion
  const int allowedChannelsSize = mAllowedChannels->size();
```

omichel commented on src/webots/nodes/WbEmitter.cpp:
```diff
@@ -105,7 +107,31 @@ void WbEmitter::updateRange() {
+bool WbEmitter::is_channel_allowed() {
+  const int allowedChannels_size = mAllowedChannels->size();
+  if (allowedChannels_size > 0) {
+    int current_channel = (int)mChannel->value();
```
```suggestion
    const int currentChannel = (int)mChannel->value();
```

omichel commented on src/webots/nodes/WbEmitter.cpp:
```diff
@@ -105,7 +107,31 @@ void WbEmitter::updateRange() {
+  if (allowedChannels_size > 0) {
+    int current_channel = (int)mChannel->value();
+    for (int i = 0; i < allowedChannels_size; i++) {
+      if (current_channel == mAllowedChannels->item(i)) {
+        return true;
+      }
```
```suggestion
      if (currentChannel == mAllowedChannels->item(i))
        return true;
```

omichel commented on src/webots/nodes/WbEmitter.cpp:
```diff
@@ -117,6 +143,10 @@ void WbEmitter::writeConfigure(QDataStream &stream) {
   stream << (double)mRange->value();
   stream << (double)mMaxRange->value();
+  stream << (int)mAllowedChannels->size();
+  for (int i = 0; i < mAllowedChannels->size(); i++) {
+    stream << (int)mAllowedChannels->item(i);
+  }
```
```suggestion
  for (int i = 0; i < mAllowedChannels->size(); i++)
    stream << (int)mAllowedChannels->item(i);
```

omichel commented on src/webots/nodes/WbEmitter.hpp:
```diff
@@ -66,17 +67,20 @@ class WbEmitter : public WbSolidDevice {
   WbEmitter &operator=(const WbEmitter &);  // non copyable
   WbNode *clone() const override { return new WbEmitter(*this); }
   void init();
+  bool is_channel_allowed();
```
```suggestion
  bool isChannelAllowed();
```

omichel commented on src/webots/nodes/WbReceiver.cpp:
```diff
@@ -177,6 +179,26 @@ void WbReceiver::updateTransmissionSetup() {
+  if (allowedChannels_size > 0) {
+    int current_channel = (int)mChannel->value();
+    for (int i = 0; i < allowedChannels_size; i++) {
+      if (current_channel == mAllowedChannels->item(i)) {
+        return true;
+      }
```
```suggestion
      if (currentChannel == mAllowedChannels->item(i))
        return true;
```

omichel commented on src/webots/nodes/WbReceiver.cpp:
```diff
@@ -177,6 +179,26 @@ void WbReceiver::updateTransmissionSetup() {
+bool WbReceiver::is_channel_allowed() {
+  const int allowedChannels_size = mAllowedChannels->size();
+  if (allowedChannels_size > 0) {
+    int current_channel = (int)mChannel->value();
```
```suggestion
    const int currentChannel = (int)mChannel->value();
```

omichel commented on src/webots/nodes/WbReceiver.cpp:
```diff
@@ -185,6 +207,10 @@ void WbReceiver::writeConfigure(QDataStream &stream) {
   stream << (int)mBufferSize->value();
   stream << (int)mChannel->value();
+  stream << (int)mAllowedChannels->size();
+  for (int i = 0; i < mAllowedChannels->size(); i++) {
+    stream << (int)mAllowedChannels->item(i);
+  }
```
```suggestion
  for (int i = 0; i < mAllowedChannels->size(); i++)
    stream << (int)mAllowedChannels->item(i);
```

omichel commented on src/webots/nodes/WbReceiver.hpp:
```diff
@@ -91,9 +93,11 @@ class WbReceiver : public WbSolidDevice {
   void receivePacketIfPossible(WbDataPacket *packet);
   bool checkApertureAndRange(const WbEmitter *emitter, const WbReceiver *receiver, bool checkRangeOnly = false) const;
   void updateRaysSetupIfNeeded() override;
+  bool is_channel_allowed();
```
```suggestion
  bool isChannelAllowed();
```

omichel reviewed (changes requested):
Great, it seems to be fully functional.
I just have a few coding style requests.

omichel commented:
> There is one small thing. What to do, when the world file contains following:
> 
> ```
> channel 1
> allowedChannels [2]
> ```
> 
> Is it possible to raise something like "error in world definition", which would prevent running the controllers? Because I'm not getting any error if I specify it in the world file.

I believe we should call `updateAllowedChannels()` in the `preFinalize()` method to address this. See for example [this](https://github.com/Adman/webots/blob/allowed-channels/src/webots/nodes/WbLens.cpp#L41-L43).

Adman commented:
@omichel Thanks. I believe I addressed all of your comments. There were a couple of more unnecessary brackets.

Also, calling `updateAllowedChannels` in `preFinalize` to print a warning.

omichel reviewed (dismissed):
Thank you for this great contribution.
Once the tests are complete, you should be able to merge this PR.

stefaniapedrazzi commented on docs/reference/emitter.md:
```diff
@@ -71,6 +72,9 @@ This is usually 8 (the default), but can be more if control bits are used.
 The total number of bytes in the packets enqueued in the emitter cannot exceed this number.
 A `bufferSize` of -1 (the default) is regarded as unlimited buffer size.
 
+- `allowedChannels`: specifies allowed channels Emitter is allowed to emit to.
```
```suggestion
- `allowedChannels`: specifies allowed channels [Emitter](#emitter) is allowed to emit to.
```

stefaniapedrazzi commented on docs/reference/receiver.md:
```diff
@@ -61,6 +62,9 @@ The noise is proportionnal to the signal strength, e.g., a `signalStrengthNoise`
 - `directionNoise`: standard deviation of the gaussian noise added to each of the components of the direction returned by `wb_receiver_get_emitter_direction`.
 The noise is not dependent on the distance between emitter-receiver.
 
+- `allowedChannels`: specifies allowed channels Receiver is allowed to listen to.
```
```suggestion
- `allowedChannels`: specifies allowed channels [Receiver](#receiver) is allowed to listen to.
```

stefaniapedrazzi commented on docs/reference/receiver.md:
```diff
@@ -641,6 +645,7 @@ channel = wb_receiver_get_channel(tag)
 *set and get the receiver's channel.*
 
 The `wb_receiver_set_channel` function allows a receiver to change its reception channel.
+Target channel must be present between allowedChannels or allowedChannels is empty.
```
```suggestion
Target channel must be present between `allowedChannels` or `allowedChannels` is empty.
```

stefaniapedrazzi reviewed (commented):
Note that it correctly prints a warning if the default channel loaded from file or set from the UI is not in the `allowedChannels` list. But it seems the invalid channel can still be used.
The `updateAllowedChanngels` should probably not only check the validity but also reset invalid values to a default value (for example the first value of the `allowedChannels` field).

stefaniapedrazzi commented on docs/reference/emitter.md:
```diff
@@ -309,6 +313,7 @@ channel = wb_emitter_get_channel(tag)
 *set and get the emitter's channel.*
 
 The `wb_emitter_set_channel` function allows the controller to change the transmission channel.
+Target channel must be present between allowedChannels or allowedChannels is empty.
```
```suggestion
Target channel must be present between `allowedChannels` or `allowedChannels` is empty.
```

stefaniapedrazzi commented on docs/reference/changelog-r2021.md:
```diff
@@ -10,6 +10,7 @@ Released on XX Xth, 2021.
     - Rework of car meshes to have more realistic rear lights for Mercedes Benz, Lincoln, Citroen, BMW and Range Rover models ([#2615](https://github.com/cyberbotics/webots/pull/2615)).
     - Add an example that shows an integration of OpenAI Gym with Webots ([#2711](https://github.com/cyberbotics/webots/pull/2711)).
     - Added a nice looking FIFA soccer ball proto ([#2782](https://github.com/cyberbotics/webots/pull/2782)).
+    - Added an allowedChannels field in the [Emitter](emitter.md) and [Receiver](receiver.md) nodes to restrict the channel usage ([#2849](https://github.com/cyberbotics/webots/pull/2849)).
```
```suggestion
    - Added an `allowedChannels` field in the [Emitter](emitter.md) and [Receiver](receiver.md) nodes to restrict the channel usage ([#2849](https://github.com/cyberbotics/webots/pull/2849)).
```

Adman commented:
@stefaniapedrazzi Yeah, going to fix it.

Adman commented:
@omichel @stefaniapedrazzi Should be fixed. After testing it I found out that after editing `allowedChannels` from the GUI it is not updated in C file, so fixed that as well. Could you have a look please? :)

Adman commented on src/controller/c/emitter.c:
```diff
@@ -159,6 +159,13 @@ static void emitter_read_answer(WbDevice *d, WbRequest *r) {
+      Emitter *es = d->pdata;
+      es->allowed_channels_size = request_read_int32(r);
+      for (int i = 0; i < es->allowed_channels_size; i++)
+        es->allowed_channels[i] = request_read_int32(r);
```
Probably need to realloc this as well..

Adman replied:
Done. Using `realloc`.

omichel reviewed (dismissed):
That looks good to me.
Thank you.

stefaniapedrazzi commented on src/webots/nodes/WbEmitter.cpp:
```diff
@@ -105,7 +109,35 @@ void WbEmitter::updateRange() {
+    parsingWarn(
+      tr("'allowedChannels' does not contain current 'channel'. Setting 'channel' to %1.").arg(mAllowedChannels->item(0)));
+    mChannel->setValue(mAllowedChannels->item(0));
+    mNeedToSetChannel = true;
```
Is this instruction really needed?
After setting the channel to the new value, the `updateChannel` is automatically called.

stefaniapedrazzi commented on src/webots/nodes/WbEmitter.cpp:
```diff
@@ -105,7 +109,35 @@ void WbEmitter::updateRange() {
 void WbEmitter::updateChannel() {
+  if (!isChannelAllowed()) {
+    parsingWarn(tr("'channel' is not included in 'allowedChannels'"));
+    return;
```
Here it is still possible to set the channel value to an invalid value because this function simply prints a warning  if the new value is invalid but doesn't reset it.

stefaniapedrazzi commented on src/webots/nodes/WbReceiver.cpp:
```diff
@@ -177,6 +180,29 @@ void WbReceiver::updateTransmissionSetup() {
   mNeedToConfigure = true;
 }
 
+bool WbReceiver::isChannelAllowed() {
```
This function should also be called in `WbReceiver::updateTransmissionSetup` to check if the new `Receiver.channel` value is valid and reset it if not.

stefaniapedrazzi reviewed (changes requested):
Thank you!

The channel value is now correctly reset if invalid after changing the `allowedChannels` values, but it is still not reset after changing `channel` value itself and it is thus still possible to use invalid channels.

omichel reviewed (approved):
To speed-up things, I allowed myself to fix the English style in the documentation.

stefaniapedrazzi reviewed (approved):
Thank you! Now it works correctly.

Adman commented:
Thank you guys!
